### ISSUE TITLE
Hybrid backed-unbacked SymInts

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#121749]
### 🐛 Describe the bug

From discussion with @suo and @avikchaudhuri 

A hybrid backed-unbacked SymInt is an unbacked SymInt that still has a hint. Specifically, when we perform `expect_true` tests on it, we generate deferred runtime asserts; only when there is an actual conditional do we generate a guard. This would help export remove "pedantic" guards that arise from error checking. Users are typically not interested in these guards, because they reflect invalid inputs on the eager model, and not potential correctness problems arising from the export process.

Related: https://github.com/pytorch/pytorch/issues/112749 where when we have real data, we can generate hybrid backed-unbacked SymInts, rather than unbacked SymInts.

### Versions

main

**Comment by avikchaudhuri** [comment]
I'm really excited about using this for export. Two reasons that conspire to make this worthwhile:
1. Currently the most complex guards come from things like reshape. They are often multivariate. In the "offline" constraint solver we use for export, these do not have a good simplification strategy, so we cop out by forcing specialization of a subset of those symbols (exactly which subset is dictated by the structure we see in examples du jour, and keeps evolving); these choices blow up later, making some users unhappy because they really wanted those symbols to be dynamic).
2. The alternative is to silently drop these complex guards *because* they cannot be expressed in the simple language of dynamic shape specifications used in export. In general this can introduce unsoundness in a way that the user can't control.

So basically by adopting hybrid symints for input shapes we can have the best of both worlds: you get guards from branchy logic which hopefully are never "too complex to specify" and you get runtime assertions for everything else that goes in the "pedantic" a.k.a. PyTorch correctness bucket. 

New issue: Can we make it a choice to remove the runtime assertions (say, via a pass) to avoid runtime cost? Or will that break the system?

**Comment by ezyang** [comment]
Whether or not runtime assertions go into the graph or not is easy to tweak. By default they're not in the graph and we have to manually insert them in.
